### PR TITLE
Add deltaiterator upperbound check

### LINKED ISSUE #2343: Transaction::GetIterator does not honor ReadOptions
The `TransactionBaseImpl::GetIterator` does not honor the `prefix_same_as_start` and `iterate_upper_bound` options for keys written in the current transaction.

The underlying `WBWIIteratorImpl` does not check these options. This means the iterator returns non-matching writes from the current transaction, which we did not expect to see based on the documentation.

This should either be documented or fixed, otherwise it leads to some frustrating bugs.

### PR DESCRIPTION
**Pull request by cz2h**
Fixes [issue 11606](https://github.com/facebook/rocksdb/issues/11606).
Fixes #2343.

Fix problem of WBWIIteratorImpl not disabling itself when it's delta iterator is at entry out of upperbound.

### CODE REVIEW

**Comment by jsteemann** [comment]
Seems to fix the issue described in https://github.com/facebook/rocksdb/issues/2343 (`iterate_upper_bound` not honored).

**Review by ajkr — APPROVED** [review]
LGTM, thanks!

**Comment by cz2h** [comment]
@ajkr @jsteemann  Many thanks for the quick review. Kindly format the code to pass the last styling test.

File: utilities/write_batch_with_index/write_batch_with_index_internal.cc [hunk 1]
```diff
@@ -307,7 +312,8 @@ void BaseDeltaIterator::UpdateCurrent() {
         if (comparator_->CompareWithoutTimestamp(
                 delta_entry.key, /*a_has_ts=*/false, *iterate_upper_bound_,
                 /*b_has_ts=*/false) >= 0) {
-          // out of upper bound -> finished.
+          // out of upper bound -> delta finished.
+          delta_iterator_out_of_bound_ = true;
```
Comment (cbi42) [inline utilities/write_batch_with_index/write_batch_with_index_internal.cc:316]: Should we reset `delta_iterator_out_of_bound_` to false when a user calls Seek* again?

**Comment by cbi42** [comment]
Another usage of `iterate_upper_bound` in db iterator is that SeekToLast() will seek to the first key before it. To make WBWI iterator consistent with base iterator behavior, we probably need to also implement that as a follow up.

File: utilities/write_batch_with_index/write_batch_with_index_internal.cc (hunk 1 above)
Comment (ajkr) [reply utilities/write_batch_with_index/write_batch_with_index_internal.cc:316]: Good point... we should probably pass the upper_bound down to the WBWIIteratorImpl and have it do the upper-bound check normally, i.e., by returning `Valid() == false` upon reaching the upper-bound.

**Comment by cz2h** [comment]
Hi @ajkr , I think the this [pr](https://github.com/facebook/rocksdb/pull/11680) provides a more detailed solution in fixing these series of issues related to the bounds of WBIIIterator and I am ok to close this one.

In the meantime could you provide some comments/thoughts on this [fix](https://github.com/facebook/rocksdb/pull/11597), thanks ?

**Pull request closed without merging**